pystoned/CNLS.py

```diff
@@ -1,304 +1,224 @@
+        ## TODO(error/warning handling): Check the configuration of the model exist
+        self.x = x.tolist()
+        self.y = y.tolist()
+        self.cet = cet
+        self.fun = fun
+        self.rts = rts
+
+        if type(self.x[0]) != list:
+            self.x = [x.tolist()]
+
+        # Initialize the pyomo model
+        self.__model__ = ConcreteModel()
+        self.__model__.I = Set(initialize=range(len(self.y)))
+        self.__model__.J = Set(initialize=range(len(self.x[0])))
+
+        # Initialize the variables
+        self.__model__.alpha = Var(self.__model__.I, doc='alpha')
+        self.__model__.beta = Var(
+            self.__model__.I, self.__model__.J, bounds=(0.0, None), doc='beta')
+        self.__model__.epsilon = Var(self.__model__.I, doc='residuals')
+        self.__model__.frontier = Var(
+            self.__model__.I, bounds=(0.0, None), doc='estimated frontier')
+
+        # Setup the objective function and constraints for model
+        self.__model__.objective = Objective(
+            rule=self.__objective_rule(),
+            sense=minimize,
+            doc='objective function')
+        self.__model__.regression_rule = Constraint(
+            self.__model__.I,
+            rule=self.__regression_rule(),
+            doc='regression equation')
+        if self.cet == "mult":
+            self.__model__.log_rule = Constraint(
+                self.__model__.I,
+                rule=self.__log_rule(),
+                doc="regression function")
+        self.__model__.concav_convex_rule = Constraint(
+            self.__model__.I,
+            self.__model__.I,
+            rule=self.__concav_convex_rule(),
+            doc='concavity/convexity constraint')
+
+        # Optimize model
+        self.optimization_status = 0
+        self.problem_status = 0
+
+    def optimize(self, remote=False):
+        """Optimize the function by requested method"""
+        ## TODO(error/warning handling): Check problem status after optimization
+        if remote == False:
+            solver = SolverFactory('mosek')
+            self.problem_status = solver.solve(self.__model__, tee=True)
+            self.optimization_status = 1
+        else:
+            solver = SolverManagerFactory('neos')
+            self.problem_status = solver.solve(
+                self.__model__, tee=True, opt='knitro')
+            self.optimization_status = 1
+
+    def __objective_rule(self):
+        """Return the proper objective function for config"""
+
+        def objective_rule(model):
+            return sum(model.epsilon[i]**2 for i in model.I)
+
+        return objective_rule
+
+    def __regression_rule(self):
+        """Return the proper regression constraint for config"""
+        if self.cet == "addi":
+            if self.rts == "vrs":
+
+                def regression_rule(model, i):
+                    return self.y[i] == model.alpha[i] \
+                        + sum(model.beta[i, j] * self.x[i][j] for j in model.J)\
```

Comment:
`sum(model.beta[i, j] * self.x[i][j] for j in model.J)`:

There is another scenario we should take into consideration.  When modeling the 1 input-1 output case, we only need to estimate one beta coefficient. That is, the `model.beta[i, j]` should be the `model.beta[i].`  That is also why I use if-condition for "m" in the original CNLS.

Reply:
This has been considered in line 22-23:
```python
if type(self.x[0]) != list:
    self.x = [x.tolist()]
```
All the input are considered 2 dimensional.
While it is 1 input, this part will transfer the 1 dimensional list to 2 dimensional list.
![image](https://user-images.githubusercontent.com/37362730/89165745-8f636080-d5ab-11ea-8b08-f1c4800b13e8.png)
Which can reduce lots of code for situation of single input.
This well applied in many situations, since 1 dimensional list is tricky in many packages.

Reply:
OK. Thanks. Good to know that.